File: leetcode/python/0424. Longest Repeating Character Replacement.py
# Time complexity: O(N)
# Space complexity: O(1)

class Solution(object):
    def characterReplacement(self, s, k):
        freq_table = {}

        left = 0
        right = 0
        res = 0

        freq_table[s[left]] = 1

        while right < len(s)-1:
            right += 1
            if s[right] in freq_table:
                freq_table[s[right]] += 1
            else:
                freq_table[s[right]] = 1
            
            if right-left+1 - self.most_frequent(freq_table) > k:
                freq_table[s[left]] -= 1
                left += 1
            res = max(res, right-left+1)  
        return res          

            
    def most_frequent(self, freq_table):
        return max(freq_table.values())

File: leetcode/java/0424. Longest Repeating Character Replacement.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(1)

class Solution {
    public int characterReplacement(String s, int k) {
        int[] characterMap = new int[26];

        int left = 0;
        int right = 0;
        int longestSub = 0;

        while (right < s.length()) {
            char currentChar = s.charAt(right);
            characterMap[(int) currentChar - (int) 'A'] += 1;

            int mostFreq = this.findMostFreq(characterMap);
            int numProcessed = right - left + 1;
            if (numProcessed - mostFreq <= k) {
                longestSub = Integer.max(numProcessed, longestSub);
            } else {
                characterMap[(int) s.charAt(left) - (int) 'A'] -= 1;
                left += 1;
            }
            right += 1;
        }
        return longestSub;
    }

    public int findMostFreq(int[] arr) {
        int mostFreq = 0;

        for (int num : arr) {
            if (num > mostFreq) {
                mostFreq = num;
            }
        }
        return mostFreq;
    }
}